Supports 'ensure exist' class addon in Addon-manager

MrHohn opened the pull request:
Fixes #39561, fixes #37047 and fixes #36411. Depends on #40057.

This PR splits cluster addons into two categories:
- Reconcile: Addons that need to be reconciled (`kube-dns` for instance).
- EnsureExists: Addons that need to be exist but changeable (`default-storage-class`).

The behavior for the 'EnsureExists' class addon would be:
- Create it if not exist.
- Users could do any modification they want, addon-manager will not reconcile it.
- If it is deleted, addon-manager will recreate it with the given template.
- It will not be updated/clobbered during upgrade.

As Brian pointed out in [#37048/comment](https://github.com/kubernetes/kubernetes/issues/37048#issuecomment-272510835), this may not be the best solution for addon-manager. Though #39561 needs to be fixed in 1.6 and we might not have enough bandwidth to do a big surgery.

@mikedanese @thockin 

cc @kubernetes/sig-cluster-lifecycle-misc 

---

Tasks for this PR:
- [x] Supports 'ensure exist' class addon and switch to use new labels in addon-manager.
- [x] Updates READMEs regarding the new behavior of addon-manager.
- [x] Updated `test/e2e/addon_update.go` to match the new behavior.
- [x] Go through all current addons and apply the new labels on them regarding what they need.
- [x] Bump addon-manager and update its template files.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/40108)
<!-- Reviewable:end -->

pires reviewed (approved):
lgtm

mikedanese commented on cluster/addons/addon-manager/kube-addons.sh:
```diff
@@ -159,6 +164,15 @@ function update_addons() {
+function ensure_addons() {
+  # Create objects already exist should fail.
+  # Filter out `AlreadyExists` message to not noisily log.
+  ${KUBECTL} ${KUBECTL_OPTS} create --namespace=${SYSTEM_NAMESPACE} -f ${ADDON_PATH} -l ${SERVICE_ENSURE_EXIST_LABEL} \
```
What is the error code of this?

MrHohn replied:
For the `kubectl` command, try to create something that already exist will exit with 255. And successfully create all objects should exit with 0. Though another common case would be some creations succeed while some creations fail. Not sure what the exit code would be in this case.

I removed the line "completed successfully" and "completed with errors" for this function because it is somehow hard to tell whether it run correctly or not.

thockin replied:
should we just patch kubectl to have a `--ignore-already-exists` flag to create, similar to `--ignore-not-found` on delete?

MrHohn replied:
Makes sense, will try to patch that.

MrHohn replied:
Also bad news on this one. I tried implement `--ignore-already-exists` for `kubectl create` but encountered another problem: `kubectl create` returns error when there is no object passed to create. With this flag enabled, if we pass in a few objects to create but all of them exist, we still get an error because the "nothing create error".

I need to put more thoughts on it. But feel like this could be done later.

mikedanese commented:
This looks good. Can you bump addon manager version in this PR ?

luxas commented:
Please a little bit with updating the tag. I have an open PR which I have rebased countless times that @ixdy said he was gonna take a look at today and maybe merge that updates this same file: https://github.com/kubernetes/kubernetes/pull/38926

MrHohn commented:
> Can you bump addon manager version in this PR ?

@mikedanese Sure, but we need to wait for #40057 get reviewed and merged. Otherwise this won't work.

thockin commented on cluster/addons/addon-manager/kube-addons.sh:
```diff
@@ -150,7 +155,7 @@ function update_addons() {
 
   run_until_success "${KUBECTL} ${KUBECTL_OPTS} apply --namespace=${SYSTEM_NAMESPACE} -f ${ADDON_PATH} \
-    --prune=${enable_prune} -l kubernetes.io/cluster-service=true --recursive ${additional_opt}" 3 5
+    --prune=${enable_prune} -l ${SERVICE_RECONCILE_LABEL} --recursive ${additional_opt}" 3 5
```
rename this function to be clearer?

MrHohn replied:
You mean `update_addons`? What about `reconcile_addons()`?

thockin commented:
naming it to match the labels seems sane.

On Sun, Jan 22, 2017 at 9:31 PM, Zihong Zheng <notifications@github.com>
wrote:

> *@MrHohn* commented on this pull request.
> ------------------------------
>
> In cluster/addons/addon-manager/kube-addons.sh
> <https://github.com/kubernetes/kubernetes/pull/40108>:
>
> > @@ -150,7 +155,7 @@ function update_addons() {
>    local -r additional_opt=$2;
>
>    run_until_success "${KUBECTL} ${KUBECTL_OPTS} apply --namespace=${SYSTEM_NAMESPACE} -f ${ADDON_PATH} \
> -    --prune=${enable_prune} -l kubernetes.io/cluster-service=true --recursive ${additional_opt}" 3 5
> +    --prune=${enable_prune} -l ${SERVICE_RECONCILE_LABEL} --recursive ${additional_opt}" 3 5
>
> You mean update_addons? What about reconcile_addons()?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/40108>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVCCKUFpvjKUtHoFSk7Mf8wglRdwjks5rVDs8gaJpZM4LndUl>
> .
>

MrHohn commented:
Updates:

As discussed offline, Addon-manager are migrating to use its own specific labels to classify addons. The new labels are:
- `addonmanager.kubernetes.io/reconcile=true`
- `addonmanager.kubernetes.io/ensure-exist=true`

Meanwhile, the old label `kubernetes.io/cluster-service` will be deprecated (only for Addon-manager). But for now, the old label will still be respected. Enumerate cases below:
- Have `kubernetes.io/cluster-service` only: being reconciled as before.
- Have `addonmanager.kubernetes.io/reconcile=true` only: being reconciled.
- Have `kubernetes.io/cluster-service` and `addonmanager.kubernetes.io/reconcile=true`: being reconciled.
- Have `kubernetes.io/cluster-service` and `addonmanager.kubernetes.io/ensure-exist=true`: ensuring it exist.
- Have `addonmanager.kubernetes.io/ensure-exist=true` only: ensuring it exist.

More detail could be found in the updated READMEs.

Tasks for this PR:
- [x] Supports 'ensure exist' class addon and switch to use new labels in Addon-manager.
- [x] Updates READMEs regarding the new behavior of Addon-manager.
- [ ] Bump Addon-manager to a v6.3 and update its template files.
- [ ] Go through all current addons and apply the new labels on them regarding what they need.
- [ ] Updated `test/e2e/addon_update.go` to match its new behavior.

Push the first two commits ahead to check it is on the right track.

MrHohn commented on cluster/addons/addon-manager/kube-addons.sh:
```diff
@@ -145,20 +155,34 @@ function annotate_addons() {
+  # ADDON_ENSURE_EXIST_LABEL will also be reconciled.
   run_until_success "${KUBECTL} ${KUBECTL_OPTS} apply --namespace=${SYSTEM_NAMESPACE} -f ${ADDON_PATH} \
-    --prune=${enable_prune} -l kubernetes.io/cluster-service=true --recursive ${additional_opt}" 3 5
+    -l ${ADDON_RECONCILE_LABEL}=true -l '${CLUSTER_SERVICE_LABEL}=true,!${ADDON_ENSURE_EXIST_LABEL}' \
```
These two `-l` mean:
```
if match ${ADDON_RECONCILE_LABEL}=true ||
    (match ${CLUSTER_SERVICE_LABEL}=true && don't have ${ADDON_ENSURE_EXIST_LABEL}) {
    reconcile_it()...
}
```

MrHohn replied:
Sorry, this usage for multiple selectors will not work. Discussion on #40588.

spxtr commented:
@k8s-bot bazel test this
#40526

mikedanese commented:
/approve

thockin commented:
Why 2 labels:

* addonmanager.kubernetes.io/reconcile=true
* addonmanager.kubernetes.io/ensure-exist=true

vs a single label and enumerated value?

* addonmanager.kubernetes.io/mode=Reconcile
* addonmanager.kubernetes.io/mode=EnsureExists



On Thu, Jan 26, 2017 at 6:34 PM, Kubernetes Submit Queue <
notifications@github.com> wrote:

> [APPROVALNOTIFIER] This PR is *APPROVED*
>
> The following people have approved this PR: *mikedanese*
>
> Needs approval from an approver in each of these OWNERS Files:
>
>    - cluster/OWNERS
>    <https://github.com/kubernetes/kubernetes/blob/master/cluster/OWNERS>
>    [mikedanese]
>
> You can indicate your approval by writing /approve in a comment
> You can cancel your approval by writing /approve cancel in a comment
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/40108#issuecomment-275575435>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVOKTAq_PzPW6GHeUMZPh65cSwIseks5rWVe0gaJpZM4LndUl>
> .
>

MrHohn commented:
@thockin Yeah single label seems better. Label value=true/false does not make much sense.

ixdy commented:
fyi https://github.com/kubernetes/kubernetes/pull/38926 bumps addon-manager to v6.3, and the image has already been pushed, though the specs/salt haven't been updated yet.

thockin commented on cluster/addons/README.md:
```diff
@@ -1,53 +1,34 @@
+by the initial config. In order to make Horizontal / Vertical Auto-scaling functional,
+the related fields in config should be left unset. More specifically, leave `replicas`
+in `ReplicationController` / `Deployment` / `ReplicaSet` unset for Horizontal Scaling,
+leave `resources` for container unset for Vertical Scaling. The periodical reconcile
```
s/periodical/periodic

thockin commented on cluster/addons/README.md:
```diff
@@ -1,53 +1,34 @@
+
+## Cooperating Horizontal / Vertical Auto-Scaling with "reconcile class addons"
+
+"reconcile class addons" will be periodically reconciled to the original state given
```
Reconcile is capitalized.

thockin replied:
I would only quote "Reconcile" here

thockin commented on cluster/addons/README.md:
```diff
@@ -1,53 +1,34 @@
+the related fields in config should be left unset. More specifically, leave `replicas`
+in `ReplicationController` / `Deployment` / `ReplicaSet` unset for Horizontal Scaling,
+leave `resources` for container unset for Vertical Scaling. The periodical reconcile
+won't clobbered these fields, hence they could be managed by Horizontal / Vertical
```
clobber

thockin commented on cluster/addons/addon-manager/README.md:
```diff
@@ -1,14 +1,35 @@
+### Addon-manager
+
+Addon-manager manages two classes of addons with given template files.
+- Addons with label `addonmanager.kubernetes.io/reconcile=true` will be periodically
```
addon-manager to match the dir name.

thockin commented on cluster/addons/addon-manager/README.md:
```diff
@@ -1,14 +1,35 @@
+
+Addon-manager manages two classes of addons with given template files.
+- Addons with label `addonmanager.kubernetes.io/reconcile=true` will be periodically
+reconciled. Direct manipulation to these addons through Apiserver is discouraged because
```
"of these addons"

apiserver or "the API server"

thockin commented on cluster/addons/addon-manager/README.md:
```diff
@@ -1,14 +1,35 @@
+Addon-manager manages two classes of addons with given template files.
+- Addons with label `addonmanager.kubernetes.io/reconcile=true` will be periodically
+reconciled. Direct manipulation to these addons through Apiserver is discouraged because
+Addon-manager will bring them back to the original state, in particular:
```
addon-manager, no caps

thockin commented on cluster/addons/addon-manager/README.md:
```diff
@@ -1,14 +1,35 @@
+	- If an addon is updated through Apiserver, it will be reconfigured to the state given
+	by the supplied fields in the template file.
+	- Corresponding addon resource will be deleted when the manifest file is deleted.
+- Addons with label `addonmanager.kubernetes.io/ensure-exist=true` will be ensured exist.
```
will be checked for existence only.

thockin commented on cluster/addons/addon-manager/README.md:
```diff
@@ -1,14 +1,35 @@
+	by the supplied fields in the template file.
+	- Corresponding addon resource will be deleted when the manifest file is deleted.
+- Addons with label `addonmanager.kubernetes.io/ensure-exist=true` will be ensured exist.
+User can edit these addons as they want. In particular:
```
Users

thockin commented on cluster/addons/addon-manager/README.md:
```diff
@@ -1,14 +1,35 @@
+- Addons with label `addonmanager.kubernetes.io/ensure-exist=true` will be ensured exist.
+User can edit these addons as they want. In particular:
+	- They will only be created/re-created with the given template files when there is no
+	such resource exist.
```
s/such resource exist/instance of the resource with that name/

thockin commented on cluster/addons/addon-manager/README.md:
```diff
@@ -1,14 +1,35 @@
+
+Notes:
+- Label `kubernetes.io/cluster-service=true` is deprecated (only for Addon Manager).
+In future release (after k8s v1.6), Addon Manager may not respect it anymore. Addons
```
Deprecation policy says 1 year.

thockin commented on cluster/addons/addon-manager/README.md:
```diff
@@ -1,14 +1,35 @@
+In future release (after k8s v1.6), Addon Manager may not respect it anymore. Addons
+with this label but without `addonmanager.kubernetes.io/ensure-exist=true` will be
+treated as "reconcile class addons" for now.
+- Resource under $ADDON_PATH (default `/etc/kubernetes/addons/`) needs to have either one
```
Resources

thockin replied:
s/needs/need

thockin commented on cluster/addons/addon-manager/README.md:
```diff
@@ -1,14 +1,35 @@
+treated as "reconcile class addons" for now.
+- Resource under $ADDON_PATH (default `/etc/kubernetes/addons/`) needs to have either one
+of these two labels (not include the deprecated label) mentioned above. Meanwhile namespaced
+resource needs to be in `kube-system` namespace. Otherwise it will be omitted. 
```
resources need

thockin replied:
Otherwise they

thockin commented on cluster/addons/addon-manager/kube-addons.sh:
```diff
@@ -37,6 +37,16 @@ ADDON_PATH=${ADDON_PATH:-/etc/kubernetes/addons}
+# also be reconciled for now.
+CLUSTER_SERVICE_LABEL="kubernetes.io/cluster-service"
+# Addon Manager reconciles addon resources with label ADDON_RECONCILE_LABEL=true.
+ADDON_RECONCILE_LABEL="addonmanager.kubernetes.io/reconcile"
```
ping me when you change this to one label (if that is still happening)

thockin reviewed (commented):
Mostly just English nits.  waiting for annotation format change

MrHohn commented:
Sorry for the confusing state here. Changed to use one label with modes. But this implementation is now buggy after I split the `kubectl apply -l XXX -l XXX`, which actually won't work, into two commands. One related issue is #40635.

Will update with more details and cases later.

MrHohn commented:
@k8s-bot cross build this

MrHohn commented:
Seems like all the puzzles have been solved. The new commits update addon manager e2e test and add  "addonmanager.kubernetes.io/mode" label to cluster addons. Right now the only "ensure exist" class addon is `default-storage-class`.

This should also fix #39561. cc @msau42

Ready for review now. @mikedanese  @thockin

MrHohn commented on cluster/saltbase/salt/kube-addons/kube-addon-manager.yaml:
```diff
@@ -12,7 +12,8 @@ spec:
     # - cluster/images/hyperkube/static-pods/addon-manager-multinode.json
-    image: gcr.io/google-containers/kube-addon-manager:v6.2
+    # image: gcr.io/google-containers/kube-addon-manager-amd64:v6.4
+    image: gcr.io/zihongz-k8s-dev/kube-addon-manager-amd64:v6.4-pr
```
Will push the new images after approved.

thockin commented:
I glanced.  It looks good overall.  I abdicate any authority I had on final review to @mikedanese :)

msau42 commented:
cc @jsafrane

MrHohn commented:
Friendly ping since it is getting close to code freeze :) @mikedanese 

(The conflicts are all version tweaks and unrelated changes so I kept them there.)

thockin commented:
@mikedanese might be busy this week - can we nominate an alternate reviewer?

MrHohn commented:
@luxas @pires @jbeda Would anyone has bandwidth to help on the final review? Sorry for getting you involved, but we really need this in for the default storage class fix :(

roberthbailey commented:
@zmerlynn would be a good reviewer candidate if he has time to take a look.

pires commented:
Unfortunately, don't have the bw to review the changes before code-freeze. It's been a while since I last looked into this.

msau42 commented:
Can someone please review this change?  We really need this in 1.6 to allow users to disable default storage classes.

thockin commented:
This LGTM.  Needs rebase

MrHohn committed:
Supports 'ensure exist' class addon in addon-manager

MrHohn committed:
Updates READMEs regarding the new behavior of addon-manager

MrHohn committed:
Updates test/e2e/addon_update.go to match addon-manager's new behavior

MrHohn committed:
Bumps addon-manager to v6.4-alpha.3 and updates template files

MrHohn committed:
Adds the new addon-manager labels on cluster addon templates

MrHohn commented:
Many thanks for taking the time! Rebased and pushed the real images:
- gcr.io/google-containers/kube-addon-manager:v6.4-alpha.3
- gcr.io/google-containers/kube-addon-manager-amd64:v6.4-alpha.3
- gcr.io/google-containers/kube-addon-manager-arm:v6.4-alpha.3
- gcr.io/google-containers/kube-addon-manager-arm64:v6.4-alpha.3
- gcr.io/google-containers/kube-addon-manager-ppc64le:v6.4-alpha.3
- gcr.io/google-containers/kube-addon-manager-s390x:v6.4-alpha.3

Also verified the slow test passed locally :)

MrHohn commented:
BTW, also worth to mention that the new images are built with kubectl binary on HEAD (17375fc59fff39135af63bd1750bb07c36ef873b) instead of the released one to unblock #41765.
